interp/iostream.go

```diff
@@ -0,0 +1,203 @@
+package interp
+
+import (
+	"bufio"
+	"io"
+	"os/exec"
+	"syscall"
+)
+
+const (
+	notClosedExitCode = -127
+)
+
+func cmdExitCode(cmdError error, ioerrs... error) (exitCode int, err error) {
+	var ee *exec.ExitError
+	var ok bool
+
+	if cmdError == nil {
+		// exitCode = 0.0
+		goto CheckIOErrors
+	}
+	if ee, ok = cmdError.(*exec.ExitError); !ok {
+		exitCode = -1.0
+		return
+	}
+	if exitCode, err = exitErrorCode(ee); err != nil {
+		return
+	}
+
+CheckIOErrors:
+	for _, err = range ioerrs {
+		if err != nil {
+			break
+		}
+	}
+	return
+}
+
+// Copy the behavior of [gm]awk for pipe exit status. n?awk returns the raw
+// wait(3) return value but this is hard to use without a bit shift operator!
+func exitErrorCode(ee *exec.ExitError) (exitCode int, err error) {
+	stat, ok := ee.ProcessState.Sys().(syscall.WaitStatus)
+	if !ok {
+		// Not all platforms support WaitStatus?
+		return -1.0, err
+	}
+	switch {
+	case stat.Exited():
+		exitCode = stat.ExitStatus()
+	case stat.Signaled():
+		exitCode = 256 + int(stat.Signal())
+		if stat.CoreDump() {
+			exitCode += 256
+		}
+	default:
+		exitCode = -1
+		err = ee
+	}
+	return
+}
+
+type inputStream interface {
+	io.ReadCloser
+	ExitCode() int
+}
+
+type outputStream interface {
+	io.WriteCloser
+	Flush() error
+	ExitCode() int
+}
+
+type outFileStream struct {
+	*bufio.Writer
+	io.Closer
```

Comment:
If you do that then you would also need to write a `Flush` method one-liner. I think it would be better to leave the `*bufio.Writer` embedded and make an implicit `io.Closer` field. I suppose it doesn't make sense to embed the `io.Closer` if the `Close() method is overridden and that is the only thing the `io.Closer` provides.


```go
type outFileStream struct {
    *bufio.Writer
    closer io.Closer
    exitCode int
}

func (s *outFileStream) Close() error { ... }
```